fpga_interchange/fasm_generators/xc7.py

```diff
@@ -142,52 +144,147 @@ def handle_luts(self):
+
+    def handle_slice_ff(self):
+        """
+        Handles slice FFs FASM feature emission.
+        """
+
+        allowed_cell_types = ["FDRE", "FDSE", "FDCE", "FDPE", "LDCE", "LDPE"]
+        allowed_site_types = ["SLICEL", "SLICEM"]
+
+        for cell_instance, cell_data in self.physical_cells_instances.items():
+            cell_type = cell_data.cell_type
+            if cell_type not in allowed_cell_types:
+                continue
+
+            site_name = cell_data.site_name
+            site_type = cell_data.site_type
+
+            if site_type not in allowed_site_types:
+                continue
+
+            tile_name = cell_data.tile_name
+            tile_type = cell_data.tile_type
+            sites_in_tile = cell_data.sites_in_tile
+            slice_site = self.get_slice_prefix(site_name, tile_type,
+                                               sites_in_tile)
+
+            bel = cell_data.bel
+
+            if cell_type in ["FDRE", "FDCE", "LDCE"]:
+                self.add_cell_feature((tile_name, slice_site, bel, "ZRST"))
+
+            if cell_type.startswith("LD"):
+                self.add_cell_feature((tile_name, slice_site, "LATCH"))
+
+            if cell_type in ["FDRE", "FDCE"]:
+                self.add_cell_feature((tile_name, slice_site, "FFSYNC"))
+
+            init_param = self.device_resources.get_parameter_definition(
+                cell_data.cell_type, "INIT")
+            init_value = init_param.decode_integer(
+                cell_data.attributes["INIT"])
+
+            if init_value == 0:
+                self.add_cell_feature((tile_name, slice_site, bel, "ZINI"))
+
+    def handle_clock_resources(self):
+        for cell_instance, cell_data in self.physical_cells_instances.items():
+            cell_type = cell_data.cell_type
+            if cell_type not in ["BUFG", "BUFGCTRL"]:
+                continue
+
+            site_name = cell_data.site_name
+            site_type = cell_data.site_type
+            site_loc = cell_data.sites_in_tile.index(site_name)
+            site_prefix = "BUFGCTRL.BUFGCTRL_X0Y{}".format(site_loc)
+
+            tile_name = cell_data.tile_name
+
+            self.add_cell_feature((tile_name, site_prefix, "IN_USE"))
+
+            if cell_type == "BUFG":
+                for feature in ["IS_IGNORE1_INVERTED", "ZINV_CE0", "ZINV_S0"]:
+                    self.add_cell_feature((tile_name, site_prefix, feature))
 
     def handle_site_thru(self, site_thru_pips):
         """
         This function is currently specialized to add very specific features
         for pseudo PIPs which need to be enabled to get the correct HW behaviour
         """
 
+        def get_feature_prefix(site_thru_feature, wire):
+            regex = re.compile(site_thru_feature.regex)
+
+            m = regex.match(wire)
+
+            return site_thru_feature.callback(m) if m else None
+
+        SiteThruFeature = namedtuple('SiteThruFeature',
```

Comment:
This type definition should be at the top of the file, along with some comments around it.

Reply:
I have merged this namedtuple with the other one (PipRegex) as they were very similar, and moved it to the top with a comment.